docs/book/how-to/handle-data-artifacts/datasets.md

```diff
@@ -0,0 +1,240 @@
+from zenml.io import fileio
+from zenml.enums import ArtifactType
+import json
+import os
+import tempfile
+import pandas as pd
+
+
+class CSVDatasetMaterializer(BaseMaterializer):
+    ASSOCIATED_TYPES = (CSVDataset,)
+    ASSOCIATED_ARTIFACT_TYPE = ArtifactType.DATA
+
+    def load(self, data_type: Type[CSVDataset]) -> CSVDataset:
+        # Create a temporary file to store the CSV data
+        with tempfile.NamedTemporaryFile(delete=False, suffix='.csv') as temp_file:
+            # Copy the CSV file from the artifact store to the temporary location
+            with fileio.open(os.path.join(self.uri, "data.csv"), "rb") as source_file:
+                temp_file.write(source_file.read())
+            
+            temp_path = temp_file.name
+
+        # Create and return the CSVDataset
+        dataset = CSVDataset(temp_path)
+        dataset.read_data()
+        return dataset
+
+    def save(self, dataset: CSVDataset) -> None:
+        # Ensure we have data to save
+        df = dataset.read_data()
+
+        # Save the dataframe to a temporary CSV file
+        with tempfile.NamedTemporaryFile(delete=False, suffix='.csv') as temp_file:
+            df.to_csv(temp_file.name, index=False)
+            temp_path = temp_file.name
+
+        # Copy the temporary file to the artifact store
+        with open(temp_path, "rb") as source_file:
+            with fileio.open(os.path.join(self.uri, "data.csv"), "wb") as target_file:
+                target_file.write(source_file.read())
+
+        # Clean up the temporary file
+        os.remove(temp_path)
+
+class BigQueryDatasetMaterializer(BaseMaterializer):
+    ASSOCIATED_TYPES = (BigQueryDataset,)
+    ASSOCIATED_ARTIFACT_TYPE = ArtifactType.DATA
+
+    def load(self, data_type: Type[BigQueryDataset]) -> BigQueryDataset:
+        with fileio.open(os.path.join(self.uri, "metadata.json"), "r") as f:
+            metadata = json.load(f)
+        dataset = BigQueryDataset(
+            table_id=metadata["table_id"],
+            project=metadata["project"],
+        )
+        dataset.read_data()
+        return dataset
+
+    def save(self, bq_dataset: BigQueryDataset) -> None:
+        metadata = {
+            "table_id": bq_dataset.table_id,
+            "project": bq_dataset.project,
+        }
+        with fileio.open(os.path.join(self.uri, "metadata.json"), "w") as f:
+            json.dump(metadata, f)
+        if bq_dataset.df is not None:
+            bq_dataset.write_data()
+```
+
+## Managing Complexity in Pipelines with Multiple Data Sources
+
+When working with multiple data sources, it's crucial to design flexible pipelines that can handle different scenarios. Here's an example of how to structure a pipeline that works with both CSV and BigQuery datasets:
+
+```python
+from zenml import step, pipeline
+from typing_extensions import Annotated
+
+@step(output_materializer=CSVDatasetMaterializer)
+def extract_data_local(data_path: str = "data/raw_data.csv") -> CSVDataset:
```

Comment:
Either the return type should be `Dataset` or the output materializer is not necessary, no?

Reply:
I wanted to be clear what materializer gets picked up... I think we need to do it?

Reply:
In this case the output is the exact class, I don't think the materializer needs to be specified.
I think what we really need is to actually run this code.